src/passkey_child/passkey_child_assert.c

```diff
@@ -0,0 +1,261 @@
+/*
+    SSSD
+
+    Helper child to commmunicate with passkey devices
+
+    Authors:
+        Iker Pedrosa <ipedrosa@redhat.com>
+
+    Copyright (C) 2022 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <termios.h>
+#include <stdio.h>
+#include <openssl/ec.h>
+#include <fido/es256.h>
+#include <fido/rs256.h>
+#include <fido/eddsa.h>
+
+#include "util/crypto/sss_crypto.h"
+#include "util/debug.h"
+#include "util/util.h"
+
+#include "passkey_child.h"
+
+errno_t
+set_assert_client_data_hash(fido_assert_t *assert)
+{
+    unsigned char cdh[32];
+    errno_t ret;
+
+    ret = sss_generate_csprng_buffer(cdh, sizeof(cdh));
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "sss_generate_csprng_buffer failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = fido_assert_set_clientdata_hash(assert, cdh, sizeof(cdh));
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "fido_assert_set_clientdata_hash failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+errno_t
+set_assert_options(const struct passkey_data *data, fido_assert_t *assert)
+{
+    errno_t ret;
+
+    ret = fido_assert_set_uv(assert, data->user_verification);
```

Comment:
In the pre-flight assertion, the device options have not yet been matched to the device. If `data->user_verification` is `FIDO_OPT_FALSE` and the authenticator does not support built-in user verification, this may cause the pre-flight to fail with `CTAP2_ERR_INVALID_OPTION` even if the authenticator knows of the credential.

Reply:
I think this is already fixed with the pre-flight assertion but take a look just in case.